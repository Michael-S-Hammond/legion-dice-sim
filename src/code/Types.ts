export enum DieColor {
    Red = "red",
    Black = "black",
    White = "white",
}

export enum AttackDieResult {
    Critical = 1,
    Hit = 2,
    Surge = 3,
    Miss = 4,
}

export type AttackRoll = {
  color: DieColor,
  result: AttackDieResult,
}

export enum AttackSurgeConversion {
  Blank = 1,
  Hit = 2,
  Critical = 3,
}

export enum Cover {
  None = "none",
  Light = "light",
  Heavy = "heavy",
}

export enum DefenseDieResult {
  Block = 1,
  Surge = 2,
  Blank = 3,
}

export type AbilityX = {
  active: boolean,
  value: number,
}

export type OffenseTokens = {
  aim: number,
  surge: number,
}

export type OffenseInput = {
  redDice: number,
  blackDice: number,
  whiteDice: number,
  surge: AttackSurgeConversion,
  tokens: OffenseTokens,
  blast: boolean,
  criticalX: AbilityX,
  duelist: boolean,
  highVelocity: boolean,
  impactX: AbilityX,
  ionX: AbilityX,
  jediHunter: boolean,
  lethalX: AbilityX,
  makashiMastery: boolean,
  pierceX: AbilityX,
  preciseX: AbilityX,
  ramX: AbilityX,
  sharpshooterX: AbilityX,
}

export type DefenseTokens = {
  dodge: number,
  observation: number,
  shield: number,
  suppression: number
  surge: number,
}

export type DefenseInput = {
  dieColor: DieColor,
  surge: boolean,
  tokens: DefenseTokens,
  cover: Cover,
  armor: boolean,
  armorX: AbilityX,
  block: boolean,
  dangerSenseX: AbilityX,
  deflect: boolean,
  djemSoMastery: boolean,
  duelist: boolean,
  hasForceUpgrades: boolean,
  immuneBlast: boolean,
  immuneMeleePierce: boolean,
  immunePierce: boolean,
  impervious: boolean,
  lowProfile: boolean,
  outmaneuver: boolean,
  soresuMastery: boolean,
  uncannyLuckX: AbilityX,
}

export type CombatInput = {
  guardian: AbilityX,
  meleeAttack: boolean,
};

export type AttackInput = {
  offense: OffenseInput,
  defense: DefenseInput,
  combat: CombatInput,
}

export type AttackResult = {
  criticals: number,
  hits: number,
  surges: number,
  misses: number,
}

export type DefenseResult = {
  forcedSaves: number,
  blocks: number,
  surges: number,
  blanks: number,
  wounds: number,
}

export type AttackOutput = {
  attack: AttackResult,
  defense: DefenseResult,
}

export type Stats = {
  mean: number,
  median: number,
  stddev: number,
}

export type AttackSummary = {
  attackCount: number,
  critical: Array<number>,
  hit: Array<number>,
  attackSurge: Array<number>,
  forcedSaves: Array<number>,
  forcedSaveStats: Stats,
  blocks: Array<number>,
  defenseSurge: Array<number>,
  wounds: Array<number>,
  woundStats: Stats
}

export type CombinedAttackOutput = {
  firstAttack: AttackOutput,
  summary: AttackSummary,
}

export enum ResultOutput {
  None = 1,
  Single = 2,
  Graph = 3,
}

export function createDefaultAttackInput() : AttackInput {
  return {
      offense: {
          redDice: 0,
          blackDice: 0,
          whiteDice: 0,
          surge: AttackSurgeConversion.Blank,
          tokens: {
            aim: 0,
            surge: 0,
          },
          blast: false,
          criticalX: { active: false, value: 1 },
          duelist: false,
          highVelocity: false,
          impactX: { active: false, value: 1 },
          ionX: { active: false, value: 1 },
          jediHunter: false,
          lethalX: { active: false, value: 1 },
          makashiMastery: false,
          pierceX: { active: false, value: 1 },
          preciseX: { active: false, value: 1 },
          ramX: { active: false, value: 1 },
          sharpshooterX: { active: false, value: 1 },
      },
      defense: {
          dieColor: DieColor.White,
          surge: true,
          tokens: {
            dodge: 0,
            observation: 0,
            shield: 0,
            suppression: 0,
            surge: 0,
          },
          cover: Cover.None,
          armor: false,
          armorX: { active: false, value: 1 },
          block: false,
          dangerSenseX: { active: false, value: 1 },
          deflect: false,
          djemSoMastery: false,
          duelist: false,
          hasForceUpgrades: false,
          immuneBlast: false,
          immuneMeleePierce: false,
          immunePierce: false,
          impervious: false,
          lowProfile: false,
          outmaneuver: false,
          soresuMastery: false,
          uncannyLuckX: { active: false, value: 1 }
      },
      combat: {
        guardian: { active: false, value: 1 },
        meleeAttack: false,
      },
  }
}

export function cloneAttackInput(original: AttackInput) : AttackInput {
  return {
      offense: {
        redDice: original.offense.redDice,
        blackDice: original.offense.blackDice,
        whiteDice: original.offense.whiteDice,
        surge: original.offense.surge,
        tokens: {
          aim: original.offense.tokens.aim,
          surge: original.offense.tokens.surge,
        },
        blast: original.offense.blast,
        criticalX: {
          active: original.offense.criticalX.active,
          value: original.offense.criticalX.value
        },
        duelist: original.offense.duelist,
        highVelocity: original.offense.highVelocity,
        impactX: {
          active: original.offense.impactX.active,
          value: original.offense.impactX.value
        },
        ionX: {
          active: original.offense.ionX.active,
          value: original.offense.ionX.value
        },
        jediHunter: original.offense.jediHunter,
        lethalX: {
          active: original.offense.lethalX.active,
          value: original.offense.lethalX.value
        },
        makashiMastery: original.offense.makashiMastery,
        pierceX: {
          active: original.offense.pierceX.active,
          value: original.offense.pierceX.value
        },
        preciseX: {
          active: original.offense.preciseX.active,
          value: original.offense.preciseX.value
        },
        ramX: {
          active: original.offense.ramX.active,
          value: original.offense.ramX.value,
        },
        sharpshooterX: {
          active: original.offense.sharpshooterX.active,
          value: original.offense.sharpshooterX.value
        },
    },
    defense: {
        dieColor: original.defense.dieColor,
        surge: original.defense.surge,
        tokens: {
          dodge: original.defense.tokens.dodge,
          observation: original.defense.tokens.observation,
          shield: original.defense.tokens.shield,
          suppression: original.defense.tokens.suppression,
          surge: original.defense.tokens.surge,
        },
        cover: original.defense.cover,
        armor: original.defense.armor,
        armorX: {
          active: original.defense.armorX.active,
          value: original.defense.armorX.value
        },
        block: original.defense.block,
        dangerSenseX: {
          active: original.defense.dangerSenseX.active,
          value: original.defense.dangerSenseX.value
        },
        deflect: original.defense.deflect,
        djemSoMastery: original.defense.djemSoMastery,
        duelist: original.defense.duelist,
        hasForceUpgrades: original.defense.hasForceUpgrades,
        immuneBlast: original.defense.immuneBlast,
        immuneMeleePierce: original.defense.immuneMeleePierce,
        immunePierce: original.defense.immunePierce,
        impervious: original.defense.impervious,
        lowProfile: original.defense.lowProfile,
        outmaneuver: original.defense.outmaneuver,
        soresuMastery: original.defense.soresuMastery,
        uncannyLuckX: {
          active: original.defense.uncannyLuckX.active,
          value: original.defense.uncannyLuckX.value
        },
    },
    combat: {
      guardian: {
        active: original.combat.guardian.active,
        value: original.combat.guardian.value
      },
      meleeAttack: original.combat.meleeAttack,
    },
  };
}
