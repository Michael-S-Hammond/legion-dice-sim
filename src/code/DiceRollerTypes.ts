import * as T from './Types';
import DiceModificationMatrix from './DiceModificationMatrix';

export type AttackResultStatus = {
    attackResult: T.AttackResult,
    modificationMatrix: DiceModificationMatrix,
    defenseResult: T.DefenseResult,
    modifiedDefenseResult: ModifiedDefenseResult,
};

type ModifiedDefenseResult = {
    blocks: number,
    surges: number,
    blanks: number,
};
