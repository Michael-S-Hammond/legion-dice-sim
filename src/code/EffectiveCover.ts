import * as T from './Types';

export function getCoverModification(cover: T.Cover) : number {
    return (cover === T.Cover.Heavy) ? 2 :
        (cover === T.Cover.Light) ? 1 :
        0;
}

export function getEffectiveCover(input: T.AttackInput) : T.Cover {
    // cover only applies to ranged attacks
    if(input.combat.meleeAttack) {
        return T.Cover.None;
    }

    // effective cover with modifications
    let effectiveCover = input.defense.cover;

    // cover from surpression
    if(input.defense.tokens.suppression > 0) {
        if(effectiveCover == T.Cover.None) {
            effectiveCover = T.Cover.Light;
        } else if(effectiveCover == T.Cover.Light) {
            effectiveCover = T.Cover.Heavy;
        }
    }

    // low profile
    if(input.defense.lowProfile && effectiveCover === T.Cover.Light) {
        effectiveCover = T.Cover.Heavy;
    }

    // sharpshooter X
    if(input.offense.sharpshooterX.active && effectiveCover !== T.Cover.None) {
        if(input.offense.sharpshooterX.value == 1) {
            if(effectiveCover === T.Cover.Heavy) {
                effectiveCover = T.Cover.Light;
            } else {
                effectiveCover = T.Cover.None;
            }
        } else if(input.offense.sharpshooterX.value == 2) {
            effectiveCover = T.Cover.None;
        }
    }

    // blast & immune: blast
    if(input.offense.blast && !input.defense.immuneBlast) {
        effectiveCover = T.Cover.None;
    }

    return effectiveCover;
}
