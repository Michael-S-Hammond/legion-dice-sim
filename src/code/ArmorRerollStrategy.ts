import * as T from "./Types";
import * as RS from "./RerollStrategy";
import * as RSH from "./RerollStrategyHelpers";

import DiceModificationMatrix, * as DMM from "./DiceModificationMatrix";
import * as EC from './EffectiveCover';

class ArmorRerollStrategy implements RS.RerollStrategy {
    pickRerollDice(input: T.AttackInput,
                   rolls: T.AttackRoll[],
                   rerollDiceCount: number,
                   reason: RS.RerollReason,
                   remaining: RS.RemainingRerolls) : number[] | undefined {
        const dmm = new DiceModificationMatrix(rolls);
        const effectiveCover = EC.getEffectiveCover(input);
        const coverModifier = EC.getCoverModification(effectiveCover);
        let rerollMiss = 0;

        // convert surges
        if(input.offense.surge === T.AttackSurgeConversion.Critical) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, rolls.length);
        }
        if(input.offense.criticalX.active) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, input.offense.criticalX.value);
        }
        if(input.offense.surge === T.AttackSurgeConversion.Hit) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Hit, rolls.length);
        }
        dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Miss, rolls.length);

        // ram
        if(input.offense.ramX.active && input.combat.meleeAttack) {
            let ramRemaining = input.offense.ramX.value;
            ramRemaining -= dmm.tryConvertResultCount(T.AttackDieResult.Miss, T.AttackDieResult.Critical, ramRemaining);
            dmm.tryConvertResultCount(T.AttackDieResult.Hit, T.AttackDieResult.Critical, ramRemaining);
        }

        // any hits to be rerolled should be converted to misses in this block
        // therefore, hits should always be 0 when this completes

        const totalHits = dmm.getResultCount(T.AttackDieResult.Hit);
        let totalCrits = dmm.getResultCount(T.AttackDieResult.Critical);
        const possibleHits = Math.min(rolls.length - totalCrits, totalHits + RSH.getMaximumRerolledDice(input, remaining));
        let hitsToKeep = 0;
        if(input.offense.impactX.active && (possibleHits > coverModifier)) {
            hitsToKeep = Math.min(totalHits, coverModifier + input.offense.impactX.value);
        }

        // remove hits for cover and armor, less impact
        dmm.tryConvertResultAllExcept(T.AttackDieResult.Hit, T.AttackDieResult.Miss, hitsToKeep);

        // calculate misses that can be rerolled
        rerollMiss = Math.min(rerollDiceCount, dmm.getResultCount(T.AttackDieResult.Miss));

        if(reason === RS.RerollReason.AimToken) {
            // convert hits to crit for impact
            const impactConvert = Math.max(0, dmm.getResultCount(T.AttackDieResult.Hit) - coverModifier);
            dmm.tryConvertResultCount(T.AttackDieResult.Hit, T.AttackDieResult.Critical, impactConvert);

            // lethal
            totalCrits = dmm.getResultCount(T.AttackDieResult.Critical);
            const effectiveLethal = RSH.getEffectiveLethal(input, 0, totalCrits, 0);
            if(effectiveLethal > 0 && remaining.aimTokens <= effectiveLethal) {
                rerollMiss = 0;
            }
        }

        // return the result
        const indexes = dmm.getRerollIndexes(0, rerollMiss);
        return indexes.length == 0 ? undefined : indexes;
    }

    shouldHandle(input: T.AttackInput, reason: RS.RerollReason) : boolean {
        return input.defense.armor;
    }
}

export default ArmorRerollStrategy;
