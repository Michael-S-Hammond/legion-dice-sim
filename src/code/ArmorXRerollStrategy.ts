import * as T from "./Types";
import * as RS from "./RerollStrategy";
import * as RSH from "./RerollStrategyHelpers";

import DiceModificationMatrix from "./DiceModificationMatrix";
import * as EC from './EffectiveCover';

class ArmorXRerollStrategy implements RS.RerollStrategy {
    pickRerollDice(input: T.AttackInput,
                   rolls: T.AttackRoll[],
                   rerollDiceCount: number,
                   reason: RS.RerollReason,
                   remaining: RS.RemainingRerolls) : number[] | undefined {
        const dmm = new DiceModificationMatrix(rolls);
        const effectiveCover = EC.getEffectiveCover(input);
        const coverModifier = EC.getCoverModification(effectiveCover);
        let rerollMiss = 0;
        let rerollHit = 0;

        // convert surges
        if(input.offense.surge === T.AttackSurgeConversion.Critical) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, rolls.length);
        }
        if(input.offense.criticalX.active) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Critical, input.offense.criticalX.value);
        }
        if(input.offense.surge === T.AttackSurgeConversion.Hit) {
            dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Hit, rolls.length);
        }
        dmm.tryConvertResultCount(T.AttackDieResult.Surge, T.AttackDieResult.Miss, rolls.length);

        rerollMiss = Math.min(rerollDiceCount, dmm.getResultCount(T.AttackDieResult.Miss));
        const startingHits = dmm.getResultCount(T.AttackDieResult.Hit);
        const totalCrits = dmm.getResultCount(T.AttackDieResult.Critical);
        const possibleHits = Math.min(rolls.length - totalCrits, startingHits + RSH.getMaximumRerolledDice(input, remaining));
        const hitsCancelledByCover = Math.min(startingHits, coverModifier);
        const hitsAfterCover = startingHits - hitsCancelledByCover;

        // convert for impact
        let effectiveImpact = 0;
        let remainingImpact = 0;
        let hitsAfterImpact = hitsAfterCover;
        if(input.offense.impactX.active) {
            effectiveImpact = Math.min(hitsAfterCover, input.offense.impactX.value);
            remainingImpact = input.offense.impactX.value - effectiveImpact;
            hitsAfterImpact = hitsAfterCover - effectiveImpact;
        }

        // convert for ram x
        let hitsAfterRam = hitsAfterImpact;
        let ramConvertMiss = 0;
        let ramConvertHit = 0;
        if(input.offense.ramX.active && input.combat.meleeAttack) {
            let remainingRam = input.offense.ramX.value;
            ramConvertMiss = Math.min(remainingRam, rerollMiss);
            rerollMiss -= ramConvertMiss;
            remainingRam -= ramConvertMiss;
            ramConvertHit = Math.min(remainingRam, hitsAfterImpact);
            hitsAfterRam -= ramConvertHit;
        }
        const effectiveRam = ramConvertMiss + ramConvertHit;

        // convert for armor x
        const remainingArmor = Math.max(0, input.defense.armorX.value - hitsAfterRam);
        const effectiveArmor = input.defense.armorX.value - remainingArmor;
        const hitsAfterArmor = hitsAfterRam - effectiveArmor;

        // determine effect of lethal
        let lethalOverride = false;
        if(reason === RS.RerollReason.AimToken) {
            const effectiveLethal = RSH.getEffectiveLethal(input, hitsAfterArmor, totalCrits + effectiveImpact, 0);
            if(effectiveLethal > 0 && remaining.aimTokens <= effectiveLethal) {
                lethalOverride = true;
            }
        }

        // it may be possible in some of the cases that rerollHit + rerollMiss > rerollDiceCount
        if(lethalOverride) {
            rerollMiss = 0;
            rerollHit = 0;
        } else if(rerollMiss === rerollDiceCount) {
            rerollHit = 0;
        } else if(possibleHits <= coverModifier) {
            rerollHit = rerollDiceCount - rerollMiss - effectiveRam;
        } else if(!input.offense.impactX.active && possibleHits <= coverModifier + input.defense.armorX.value) {
            rerollHit = rerollDiceCount - rerollMiss - effectiveRam;
        } else if(input.offense.impactX.active && possibleHits <= coverModifier + input.offense.impactX.value) {
            rerollHit = 0;
        } else if(input.offense.impactX.active && possibleHits <= coverModifier + input.offense.impactX.value + input.defense.armorX.value) {
            rerollHit = possibleHits - (coverModifier + input.offense.impactX.value) - effectiveRam;
        } else { // enough dice to get past the armor and hits to count as damage
            rerollHit = 0;
        }

        // fix possibility of rolling invalid number of dice
        if(rerollHit < 0) {
            rerollHit = 0;
        }
        if(rerollHit + rerollMiss > rerollDiceCount) {
            rerollHit = rerollDiceCount - rerollMiss;
        }

        // return the result
        const indexes = dmm.getRerollIndexes(rerollHit, rerollMiss);
        return indexes.length == 0 ? undefined : indexes;
    }

    shouldHandle(input: T.AttackInput, reason: RS.RerollReason) : boolean {
        return input.defense.armorX.active;
    }
}

export default ArmorXRerollStrategy;
