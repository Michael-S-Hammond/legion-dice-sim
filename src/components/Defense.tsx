import React from 'react';

import * as T from '../code/Types';
import * as AS from './AppStateManager';

import DefenseAbilities from './DefenseAbilities';
import DefenseTokens from './DefenseTokens';
import LoadProfileButton from './profiles/LoadProfileButton';
import SurgeDieSelector from './SurgeDieSelector';

type DefenseProps = {
    profileDialogId: string,
    showSimpleView: boolean,
    input: T.DefenseInput,
    eventHandlers: AS.AppStateDefenseEventHandlers,
}

class Defense extends React.Component<DefenseProps> {
    constructor(props : DefenseProps) {
        super(props);
    }

    private handleCoverChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
        const newCover = Number(e.target.value);
        this.props.eventHandlers.handleCoverChange(newCover);
    }

    render() : JSX.Element {
        return (
            <div>
                <h2 className="d-flex justify-content-center my-2">Defense
                    <LoadProfileButton
                        dialogId={this.props.profileDialogId}
                        tooltip='Load defense profile'></LoadProfileButton>
                </h2>
                <div className="d-flex justify-content-center my-2">
                    <SurgeDieSelector
                        color={this.props.input.dieColor}
                        surge={this.props.input.surge}
                        onClick={this.props.eventHandlers.changeDie}></SurgeDieSelector>
                </div>
                <DefenseTokens
                    showSimplifiedView={this.props.showSimpleView}
                    tokens={this.props.input.tokens}
                    eventHandlers={this.props.eventHandlers}
                    ></DefenseTokens>
                <div className="d-flex justify-content-center my-2">
                    <span className="mx-2 my-auto drop-down-label">Cover:</span>
                    <select value={this.props.input.cover} className="rounded-lg mr-4 px-2"
                        onChange={this.handleCoverChange}>
                        <option value="1">None</option>
                        <option value="2">Light</option>
                        <option value="3">Heavy</option>
                    </select>
                </div>
                <DefenseAbilities
                    showSimpleView={this.props.showSimpleView}
                    inputs={this.props.input}
                    eventHandlers={this.props.eventHandlers}
                    ></DefenseAbilities>
            </div>
        );
    }
}

export default Defense;
