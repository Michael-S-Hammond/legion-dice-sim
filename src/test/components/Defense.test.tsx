import React from 'react';

import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import * as T from '../../code/Types';
import Defense from '../../components/Defense';

import * as EventMocks from '../mocks/EventHandlerMocks';

describe('Defense', () => {
    it('matches the snapshot', () => {
        const input = T.createDefaultAttackInput();
        const events = EventMocks.createMockAppStateDefenseEventHandlers();

        const defense = shallow(<Defense
            showSimpleView={false}
            input={input.defense}
            eventHandlers={events}
            ></Defense>);

        const snapshot = shallowToJson(defense);
        expect(snapshot).toMatchSnapshot();
    });

    it('matches the snapshot for simplified view', () => {
        const input = T.createDefaultAttackInput();
        const events = EventMocks.createMockAppStateDefenseEventHandlers();

        const defense = shallow(<Defense
            showSimpleView={true}
            input={input.defense}
            eventHandlers={events}
            ></Defense>);

        const snapshot = shallowToJson(defense);
        expect(snapshot).toMatchSnapshot();
    });

    it('handles surge conversion changing', () => {
        const input = T.createDefaultAttackInput();
        const events = EventMocks.createMockAppStateDefenseEventHandlers();

        const wrapper = shallow(<Defense
            showSimpleView={false}
            input={input.defense}
            eventHandlers={events}
            ></Defense>);
        wrapper.find('select').simulate('change', { target: { value: 3 } });

        expect(events.handleCoverChange).toHaveBeenCalledTimes(1);
        expect(events.handleCoverChange).toHaveBeenCalledWith(3);
    })
});
