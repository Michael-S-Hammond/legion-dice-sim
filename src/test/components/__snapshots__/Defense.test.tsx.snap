// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Defense matches the snapshot 1`] = `
<div>
  <h2
    className="d-flex justify-content-center my-2"
  >
    Defense
    <LoadProfileButton
      dialogId="defenseProfileDialog"
      tooltip="Load defense profile"
    />
  </h2>
  <div
    className="d-flex justify-content-center my-2"
  >
    <SurgeDieSelector
      color="white"
      onClick={[MockFunction]}
      surge={true}
    />
  </div>
  <DefenseTokens
    eventHandlers={
      Object {
        "changeDie": [MockFunction],
        "handleArmorChange": [MockFunction],
        "handleArmorXChange": [MockFunction],
        "handleArmorXValueChange": [MockFunction],
        "handleBlockChange": [MockFunction],
        "handleCoverChange": [MockFunction],
        "handleDangerSenseXChange": [MockFunction],
        "handleDangerSenseXValueChange": [MockFunction],
        "handleDeflectChange": [MockFunction],
        "handleDjemSoMasteryChange": [MockFunction],
        "handleDuelistChange": [MockFunction],
        "handleHasForceUpgradesChange": [MockFunction],
        "handleImmuneBlastChange": [MockFunction],
        "handleImmunePierceChange": [MockFunction],
        "handleImperviousChange": [MockFunction],
        "handleLowProfileChange": [MockFunction],
        "handleOutmaneuverChange": [MockFunction],
        "handleSoresuMasteryChange": [MockFunction],
        "handleUncannyLuckXChange": [MockFunction],
        "handleUncannyLuckXValueChange": [MockFunction],
        "incrementDodgeTokenCount": [MockFunction],
        "incrementObservationTokenCount": [MockFunction],
        "incrementShieldTokenCount": [MockFunction],
        "incrementSuppressionTokenCount": [MockFunction],
        "incrementSurgeTokenCount": [MockFunction],
        "resetTokenCount": [MockFunction],
      }
    }
    showSimplifiedView={false}
    tokens={
      Object {
        "dodge": 0,
        "observation": 0,
        "shield": 0,
        "suppression": 0,
        "surge": 0,
      }
    }
  />
  <div
    className="d-flex justify-content-center my-2"
  >
    <span
      className="mx-2 my-auto drop-down-label"
    >
      Cover:
    </span>
    <select
      className="rounded-lg mr-4 px-2"
      onChange={[Function]}
      value="none"
    >
      <option
        value="none"
      >
        None
      </option>
      <option
        value="light"
      >
        Light
      </option>
      <option
        value="heavy"
      >
        Heavy
      </option>
    </select>
  </div>
  <DefenseAbilities
    eventHandlers={
      Object {
        "changeDie": [MockFunction],
        "handleArmorChange": [MockFunction],
        "handleArmorXChange": [MockFunction],
        "handleArmorXValueChange": [MockFunction],
        "handleBlockChange": [MockFunction],
        "handleCoverChange": [MockFunction],
        "handleDangerSenseXChange": [MockFunction],
        "handleDangerSenseXValueChange": [MockFunction],
        "handleDeflectChange": [MockFunction],
        "handleDjemSoMasteryChange": [MockFunction],
        "handleDuelistChange": [MockFunction],
        "handleHasForceUpgradesChange": [MockFunction],
        "handleImmuneBlastChange": [MockFunction],
        "handleImmunePierceChange": [MockFunction],
        "handleImperviousChange": [MockFunction],
        "handleLowProfileChange": [MockFunction],
        "handleOutmaneuverChange": [MockFunction],
        "handleSoresuMasteryChange": [MockFunction],
        "handleUncannyLuckXChange": [MockFunction],
        "handleUncannyLuckXValueChange": [MockFunction],
        "incrementDodgeTokenCount": [MockFunction],
        "incrementObservationTokenCount": [MockFunction],
        "incrementShieldTokenCount": [MockFunction],
        "incrementSuppressionTokenCount": [MockFunction],
        "incrementSurgeTokenCount": [MockFunction],
        "resetTokenCount": [MockFunction],
      }
    }
    inputs={
      Object {
        "armor": false,
        "armorX": Object {
          "active": false,
          "value": 1,
        },
        "block": false,
        "cover": "none",
        "dangerSenseX": Object {
          "active": false,
          "value": 1,
        },
        "deflect": false,
        "dieColor": "white",
        "djemSoMastery": false,
        "duelist": false,
        "hasForceUpgrades": false,
        "immuneBlast": false,
        "immunePierce": false,
        "impervious": false,
        "lowProfile": false,
        "outmaneuver": false,
        "soresuMastery": false,
        "surge": true,
        "tokens": Object {
          "dodge": 0,
          "observation": 0,
          "shield": 0,
          "suppression": 0,
          "surge": 0,
        },
        "uncannyLuckX": Object {
          "active": false,
          "value": 1,
        },
      }
    }
    showSimpleView={false}
  />
</div>
`;

exports[`Defense matches the snapshot for simplified view 1`] = `
<div>
  <h2
    className="d-flex justify-content-center my-2"
  >
    Defense
    <LoadProfileButton
      dialogId="defenseProfileDialog"
      tooltip="Load defense profile"
    />
  </h2>
  <div
    className="d-flex justify-content-center my-2"
  >
    <SurgeDieSelector
      color="white"
      onClick={[MockFunction]}
      surge={true}
    />
  </div>
  <DefenseTokens
    eventHandlers={
      Object {
        "changeDie": [MockFunction],
        "handleArmorChange": [MockFunction],
        "handleArmorXChange": [MockFunction],
        "handleArmorXValueChange": [MockFunction],
        "handleBlockChange": [MockFunction],
        "handleCoverChange": [MockFunction],
        "handleDangerSenseXChange": [MockFunction],
        "handleDangerSenseXValueChange": [MockFunction],
        "handleDeflectChange": [MockFunction],
        "handleDjemSoMasteryChange": [MockFunction],
        "handleDuelistChange": [MockFunction],
        "handleHasForceUpgradesChange": [MockFunction],
        "handleImmuneBlastChange": [MockFunction],
        "handleImmunePierceChange": [MockFunction],
        "handleImperviousChange": [MockFunction],
        "handleLowProfileChange": [MockFunction],
        "handleOutmaneuverChange": [MockFunction],
        "handleSoresuMasteryChange": [MockFunction],
        "handleUncannyLuckXChange": [MockFunction],
        "handleUncannyLuckXValueChange": [MockFunction],
        "incrementDodgeTokenCount": [MockFunction],
        "incrementObservationTokenCount": [MockFunction],
        "incrementShieldTokenCount": [MockFunction],
        "incrementSuppressionTokenCount": [MockFunction],
        "incrementSurgeTokenCount": [MockFunction],
        "resetTokenCount": [MockFunction],
      }
    }
    showSimplifiedView={true}
    tokens={
      Object {
        "dodge": 0,
        "observation": 0,
        "shield": 0,
        "suppression": 0,
        "surge": 0,
      }
    }
  />
  <div
    className="d-flex justify-content-center my-2"
  >
    <span
      className="mx-2 my-auto drop-down-label"
    >
      Cover:
    </span>
    <select
      className="rounded-lg mr-4 px-2"
      onChange={[Function]}
      value="none"
    >
      <option
        value="none"
      >
        None
      </option>
      <option
        value="light"
      >
        Light
      </option>
      <option
        value="heavy"
      >
        Heavy
      </option>
    </select>
  </div>
  <DefenseAbilities
    eventHandlers={
      Object {
        "changeDie": [MockFunction],
        "handleArmorChange": [MockFunction],
        "handleArmorXChange": [MockFunction],
        "handleArmorXValueChange": [MockFunction],
        "handleBlockChange": [MockFunction],
        "handleCoverChange": [MockFunction],
        "handleDangerSenseXChange": [MockFunction],
        "handleDangerSenseXValueChange": [MockFunction],
        "handleDeflectChange": [MockFunction],
        "handleDjemSoMasteryChange": [MockFunction],
        "handleDuelistChange": [MockFunction],
        "handleHasForceUpgradesChange": [MockFunction],
        "handleImmuneBlastChange": [MockFunction],
        "handleImmunePierceChange": [MockFunction],
        "handleImperviousChange": [MockFunction],
        "handleLowProfileChange": [MockFunction],
        "handleOutmaneuverChange": [MockFunction],
        "handleSoresuMasteryChange": [MockFunction],
        "handleUncannyLuckXChange": [MockFunction],
        "handleUncannyLuckXValueChange": [MockFunction],
        "incrementDodgeTokenCount": [MockFunction],
        "incrementObservationTokenCount": [MockFunction],
        "incrementShieldTokenCount": [MockFunction],
        "incrementSuppressionTokenCount": [MockFunction],
        "incrementSurgeTokenCount": [MockFunction],
        "resetTokenCount": [MockFunction],
      }
    }
    inputs={
      Object {
        "armor": false,
        "armorX": Object {
          "active": false,
          "value": 1,
        },
        "block": false,
        "cover": "none",
        "dangerSenseX": Object {
          "active": false,
          "value": 1,
        },
        "deflect": false,
        "dieColor": "white",
        "djemSoMastery": false,
        "duelist": false,
        "hasForceUpgrades": false,
        "immuneBlast": false,
        "immunePierce": false,
        "impervious": false,
        "lowProfile": false,
        "outmaneuver": false,
        "soresuMastery": false,
        "surge": true,
        "tokens": Object {
          "dodge": 0,
          "observation": 0,
          "shield": 0,
          "suppression": 0,
          "surge": 0,
        },
        "uncannyLuckX": Object {
          "active": false,
          "value": 1,
        },
      }
    }
    showSimpleView={true}
  />
</div>
`;
